go-controller/pkg/cni/udn/primary_network.go

```diff
@@ -2,26 +2,47 @@ package udn
 
 import (
 	"fmt"
+	"strings"
 
+	nadapi "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+	nadv1Listers "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/listers/k8s.cni.cncf.io/v1"
+	nadutils "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/utils"
+
+	corev1 "k8s.io/api/core/v1"
 	"k8s.io/klog/v2"
 
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/networkmanager"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
 )
 
-// wait on a certain pod annotation related condition
-type podAnnotWaitCond = func(map[string]string, string) (*util.PodAnnotation, bool)
+// wait on a certain pod annotation related condition. in case of non-nil error,
+// return error to abort the retry attempt
+type podAnnotWaitCond = func(*corev1.Pod, string) (*util.PodAnnotation, bool, error)
+
+const resourceNameAnnot = "k8s.v1.cni.cncf.io/resourceName"
 
 type UserDefinedPrimaryNetwork struct {
 	networkManager networkmanager.Interface
+	nadLister      nadv1Listers.NetworkAttachmentDefinitionLister
 	annotation     *util.PodAnnotation
-	activeNetwork  util.NetInfo
+
+	// the pod's UDN primary network, and its associated NAD name and the resourceName defined in the NAD
+	activeNetwork util.NetInfo
+	nadName       string
+	resourceName  string
+
+	// the pod's UDN primary interface's device information (when UDN's resourceName is defined)
+	deviceInfo *nadapi.DeviceInfo
```

Comment:
i think it would be nice to put comments on each of these variables. It will help out folks who are unfamiliar with hardware offload semantics. Like
```
// deviceInfo is extra device information stored by SRIOV device plugin in /var/run/k8s.cni.cncf.io/devinfo/dp
deviceInfo
// deviceID is either a PCI address or an Auxilary address (format <driver>.<type>.<id>, i.e. mlx5_core.sf.2) of the VF assigned to the pod
deviceID
// netdevName is the network device name in linux (such as enps0) when the VF device is bound to a kernel driver
netdevName
// isVFIO determines if the VF device is bound to the VFIO driver, used in userspace applications like DPDK.
isVFIO
```

Reply:
done